src/test/clojure/xtdb/buffer_pool_test.clj

```diff
@@ -254,3 +255,45 @@
         (util/with-open [buf @(.getBuffer bp (util/->path "aw"))]
           (let [{:keys [root]} (util/read-arrow-buf buf)]
             (util/close root)))))))
+
+(defn file-count [^Path dir]
+  (let [files (filter #(.isFile ^File %) (file-seq (.toFile dir)))]
+    (count files)))
+
+(defn insert-utf8-to-local-cache [^IBufferPool bp k len]
+  ;; PutObject on ObjectStore
+  @(.putObject bp k (utf8-buf (apply str (repeat len "a"))))
+  ;; Add to local disk cache
+  (with-open [^ArrowBuf _buf @(.getBuffer bp k)]))
+
+;; To keep in mind here, behaviour wise:
+;; - A file in considered pinned while there is a buffer in use/in the memory cache
+;; - When a file is pinned, it's weight is set to 0
+;; - When a file is unpinned, it's weight is set to the size of the file
+;; - Will explain some of the behaviour in the test below
+(t/deftest local-disk-cache-max-size
+  (util/with-tmp-dirs #{local-disk-cache}
+    (with-open [bp (bp/open-remote-storage
+                    tu/*allocator*
+                    (-> (Storage/remoteStorage simulated-obj-store-factory local-disk-cache)
+                        (.maxLocalDiskCacheBytes 10)
+                        (.maxCacheEntries 3)))]
+      (t/testing "staying below max size - all elements available"
+        (insert-utf8-to-local-cache bp (util/->path "a") 4)
+        (insert-utf8-to-local-cache bp (util/->path "b") 4)
+        (t/is (= 2 (file-count local-disk-cache)))) 
+      
+      (t/testing "going above max size - all entries pinned (ie, in memory cache) - should return all elements"
+        (insert-utf8-to-local-cache bp (util/->path "c") 4)
+        (t/is (= 3 (file-count local-disk-cache)))) 
+      
+      (t/testing "entries unpinned (cleared from memory cache by new entries) - should evict entries"
+        (insert-utf8-to-local-cache bp (util/->path "d") 4)
```

Comment:
Need to insert a number of entries to start shifting things off of the memory cache/"release" them in order for them to be considered evictable, and similarly we need to insert multiple since all pinned entries have a weight of `0` (ie, need to insert 'enough' for the weight to go over our configured max size). 

This did also initially look at what filenames were in the cache but due to the semantics of Caffeine (ie, Window TinyLfu caching rather than LRU) it's a bit unpredictable as to what exactly gets removed.

Reply:
We actually inject a bit of randomness into TinyLFU to ensure it is unpredictable in order to protect against HashDoS attacks. That's because we don't want the an artificially inflated popularity score for the LRU's victim (least recent) to block the addition of a very popular new arrival (candidate). Nor do we want to always admit or else we lose scan resistance (e.g. a multi-pass analysis loop). So if both options are popular, we accept the candidate 1% of the time to break free we'd suffer from a degraded state. This makes testing the policy harder, so in our own tests we do an ugly hack to set the thread-local random seed it a known value. I wouldn't go that far in user code since we might adjust the policy, e.g. we added hill climbing to adapt it, so tests should be behavioral making your approach nicer.

Reply:
@ben-manes Thank you for the responses on here! Some very useful suggestions and background for us.

Will probably take a similar approach here as you do in your tests :)